python/src/iceberg/io/boto.py

```diff
@@ -0,0 +1,298 @@
+            range_header = f"bytes={self._position}-"
+            self.seek(offset=0, whence=SEEK_END)
+        else:
+            position_new = self._position + n
+
+            if (
+                position_new >= self._s3_object.content_length
+            ):  # If more bytes are requested than exists, just read the entire file from the current position
+                return self.read(n=-1)
+
+            range_header = f"bytes={self._position}-{position_new -1}"
+            self.seek(offset=n, whence=SEEK_CUR)
+
+        return self._s3_object.get(Range=range_header)["Body"].read()
+
+    def seek(self, offset: int, whence: int = 0) -> int:
+        if self._closed:
+            raise ValueError("Cannot seek, file closed")
+
+        if whence not in {SEEK_SET, SEEK_CUR, SEEK_END}:
+            raise ValueError(f"Cannot seek to position {offset}, invalid whence: {whence}")
+
+        if whence == SEEK_SET:
+            self._position = offset
+        elif whence == SEEK_CUR:
+            self._position = self._position + offset
+        elif whence == SEEK_END:
+            self._position = self._s3_object.content_length - offset
+
+        return self._position
+
+    def tell(self) -> int:
+        return self._position
+
+    def closed(self) -> bool:
+        return self._closed
+
+    def close(self) -> None:
+        self._closed = True
+
+
+class BotoOutputStream(OutputStream):
+    """A wrapper for writing an S3 Object that abides by the OutputStream protocol
+
+    Args:
+    s3_object(boto3.resources.factory.s3.Object): An s3 object
+    """
+
+    def __init__(self, s3_object):
+        self._s3_object = s3_object
+        self._content = bytes()
+        self._closed = False
+
+    def write(self, b: bytes) -> None:
+        """Write to the S3 Object
+
+        Args:
+            b(bytes): The bytes to write to the S3 Object
+        """
+        if self._closed:
+            raise ValueError("Cannot write bytes, file closed")
+        self._content += b
+
+    def closed(self) -> bool:
+        """Returns whether the stream is closed or not"""
+        return self._closed
+
+    def close(self) -> None:
+        """Closes the stream and uploads the bytes to S3"""
+        self._s3_object.put(Body=self._content)
+        self._closed = True
+
+
+class BotoInputFile(InputFile):
+    """An input file implementation for the BotoFileIO
+
+    Args:
+        location(str): An S3 URI
+
+    Attributes:
+        location(str): An S3 URI
+    """
+
+    def __init__(self, location: str, session: Session, **client_kwargs):
+        parsed_location = urlparse(location)  # Create a ParseResult from the uri
+
+        if not parsed_location.scheme.startswith("s3"):
+            raise ValueError(f"Cannot create BotoInputFile, scheme not supported: {parsed_location.scheme}")
+
+        self._bucket = parsed_location.netloc
+        self._path = parsed_location.path.strip("/")
```

Comment:
This alters the path by removing any trailing `/` as well. Since this is an object and not a file system path, we don't want to do that (in S3, `a/` and `a` are distinct). Instead, I think you want to check for a starting `/` and remove only that:

```python
self.path = parsed_location.path[1:] if parsed_location.path.startswith('/') else parsed_location.path
```